sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/CompactBlockProcessor.kt

```diff
@@ -998,53 +1008,86 @@ class CompactBlockProcessor internal constructor(
+                )
+
+                Twig.debug { "Done enhancing transaction (id:${transaction.id} block:${transaction.minedHeight})" }
                 BlockProcessingResult.Success
+            } catch (e: CompactBlockProcessorException.EnhanceTransactionError) {
+                BlockProcessingResult.FailedEnhance(e)
             }
         }
 
-        private suspend fun enhanceHelper(
+        @Throws(EnhanceTxDownloadError::class)
+        private suspend fun fetchTransaction(
             id: Long,
             rawTransactionId: ByteArray,
             minedHeight: BlockHeight,
-            backend: Backend,
             downloader: CompactBlockDownloader
-        ): BlockProcessingResult {
-            Twig.debug { "START: enhancing transaction (id:$id  block:$minedHeight)" }
-
-            val fetchResponse = downloader.fetchTransaction(rawTransactionId)
-            val enhancingResult = when (fetchResponse) {
-                is Response.Success -> {
-                    @Suppress("TooGenericExceptionCaught") // RuntimeException comes from the Rust layer
-                    try {
-                        Twig.debug { "Decrypting and storing transaction (id:$id  block:$minedHeight)" }
-                        backend.decryptAndStoreTransaction(fetchResponse.result.data)
-                        Twig.debug { "DONE: enhancing transaction (id:$id  block:$minedHeight)" }
-                        BlockProcessingResult.Success
-                    } catch (exception: RuntimeException) {
-                        BlockProcessingResult.FailedEnhance(EnhanceTxDecryptError(minedHeight, exception))
+        ): ByteArray {
+            var transactionDataResult: ByteArray? = null
+            retryUpTo(TRANSACTION_FETCH_RETRIES) { failedAttempts ->
+                if (failedAttempts == 0) {
+                    Twig.debug { "Starting to fetch transaction (id:$id, block:$minedHeight)" }
+                } else {
+                    Twig.warn {
+                        "Retrying to fetch transaction (id:$id, block:$minedHeight) after $failedAttempts " +
+                            "failure(s)..."
                     }
                 }
-                is Response.Failure -> {
-                    BlockProcessingResult.FailedEnhance(
-                        EnhanceTxDownloadError(minedHeight, fetchResponse.toThrowable())
-                    )
+                when (val response = downloader.fetchTransaction(rawTransactionId)) {
+                    is Response.Success -> {
+                        transactionDataResult = response.result.data
+                    }
+                    is Response.Failure -> {
+                        throw EnhanceTxDownloadError(minedHeight, response.toThrowable())
+                    }
                 }
             }
-            return enhancingResult
+            // Result is fetched or EnhanceTxDownloadError is thrown after all attempts failed at this point
+            return transactionDataResult!!
```

Comment:
I think this could be moved into the `when` statement to eliminate the `!!`

The compiler should notice that the when statement either returns successfully or throws `EnhanceTxDownloadError`

Reply:
The problem there is the `retryUpTo` which does not allow me to return a value from it. Leaving it as it is.